In [1]:
ls

17Nov_FILES_metal_searcher_results.txt  metal_list.txt                          metals_list.txt
17Nov_metal_searcher_analysis.ipynb     metal_searcher_results.txt              test.png


In [2]:
results = open('metal_searcher_results.txt')

In [3]:
results_string = results.read()

In [4]:
results_string

"{'one_atom_residue_names': ['MG', 'MG', 'MG', 'MG', 'MG', 'MG', 'CL', 'CL', 'MG', 'CL', 'CL', 'MG', 'CL', 'CL', 'MG', 'MG', 'CL', 'CL', 'MG', 'MG', 'MG', 'MG', 'MG', 'MG', 'CL', 'CL', 'CL', 'CL', 'CL', 'CL', 'MG', 'MN', 'MN', 'MN', 'MN', 'MN', 'MN', 'MN', 'MN', 'MG', 'CL', 'MG', 'CL', 'CL', 'MG', 'CL', 'CL', 'MG', 'MG', 'MG', 'MG', 'MG', 'MG', 'CL', 'CL', 'CL', 'MG', 'MG', 'MG', 'MG', 'MG', 'CL', 'CL', 'CL', 'CL', 'CL', 'CL', 'CA', 'CL', 'CL', 'CL', 'CL', 'CL', 'NA', 'NA', 'CL', 'CL', 'NA', 'CL', 'MG', 'MG', 'CL', 'CL', 'CA', 'CL', 'CL', 'CA', 'NA', 'SR', 'CL', 'CL', 'MG', 'MG', 'MG', 'MG', 'MG', 'MG', 'MG', 'CL', 'CL', 'MG', 'NA', 'CL', 'CL', 'CL', 'CL', 'BA', 'CL', 'CL', 'CL', 'CL', 'CL', 'CL', 'CL', 'CL', 'MG', 'MG', 'MG', 'CL', 'CL', 'NA', 'NA', 'CL', 'CL', 'NI', 'NI', 'NI', 'NI', 'NI', 'NI', 'NI', 'NI', 'NI', 'CA', 'CL', 'CL', 'MG', 'CL', 'CL', 'CL', 'CL', 'MG', 'CL', 'CL', 'CL', 'CL', 'CL', 'CL', 'MG', 'MG', 'MG', 'MG', 'CL', 'CL', 'ZN', 'ZN', 'CA', 'ZN', 'ZN', 'CA', 'MG', 'CA',

In [5]:
results_dict = eval(results_string)

In [6]:
results_dict

{'error_file_count': 195,
 'ok_file_count': 113129,
 'one_atom_residue_names': ['MG',
  'MG',
  'MG',
  'MG',
  'MG',
  'MG',
  'CL',
  'CL',
  'MG',
  'CL',
  'CL',
  'MG',
  'CL',
  'CL',
  'MG',
  'MG',
  'CL',
  'CL',
  'MG',
  'MG',
  'MG',
  'MG',
  'MG',
  'MG',
  'CL',
  'CL',
  'CL',
  'CL',
  'CL',
  'CL',
  'MG',
  'MN',
  'MN',
  'MN',
  'MN',
  'MN',
  'MN',
  'MN',
  'MN',
  'MG',
  'CL',
  'MG',
  'CL',
  'CL',
  'MG',
  'CL',
  'CL',
  'MG',
  'MG',
  'MG',
  'MG',
  'MG',
  'MG',
  'CL',
  'CL',
  'CL',
  'MG',
  'MG',
  'MG',
  'MG',
  'MG',
  'CL',
  'CL',
  'CL',
  'CL',
  'CL',
  'CL',
  'CA',
  'CL',
  'CL',
  'CL',
  'CL',
  'CL',
  'NA',
  'NA',
  'CL',
  'CL',
  'NA',
  'CL',
  'MG',
  'MG',
  'CL',
  'CL',
  'CA',
  'CL',
  'CL',
  'CA',
  'NA',
  'SR',
  'CL',
  'CL',
  'MG',
  'MG',
  'MG',
  'MG',
  'MG',
  'MG',
  'MG',
  'CL',
  'CL',
  'MG',
  'NA',
  'CL',
  'CL',
  'CL',
  'CL',
  'BA',
  'CL',
  'CL',
  'CL',
  'CL',
  'CL',
  'CL',
  'CL',
  'CL',
  

In [7]:
residues = results_dict['one_atom_residue_names']

In [8]:
residues

['MG',
 'MG',
 'MG',
 'MG',
 'MG',
 'MG',
 'CL',
 'CL',
 'MG',
 'CL',
 'CL',
 'MG',
 'CL',
 'CL',
 'MG',
 'MG',
 'CL',
 'CL',
 'MG',
 'MG',
 'MG',
 'MG',
 'MG',
 'MG',
 'CL',
 'CL',
 'CL',
 'CL',
 'CL',
 'CL',
 'MG',
 'MN',
 'MN',
 'MN',
 'MN',
 'MN',
 'MN',
 'MN',
 'MN',
 'MG',
 'CL',
 'MG',
 'CL',
 'CL',
 'MG',
 'CL',
 'CL',
 'MG',
 'MG',
 'MG',
 'MG',
 'MG',
 'MG',
 'CL',
 'CL',
 'CL',
 'MG',
 'MG',
 'MG',
 'MG',
 'MG',
 'CL',
 'CL',
 'CL',
 'CL',
 'CL',
 'CL',
 'CA',
 'CL',
 'CL',
 'CL',
 'CL',
 'CL',
 'NA',
 'NA',
 'CL',
 'CL',
 'NA',
 'CL',
 'MG',
 'MG',
 'CL',
 'CL',
 'CA',
 'CL',
 'CL',
 'CA',
 'NA',
 'SR',
 'CL',
 'CL',
 'MG',
 'MG',
 'MG',
 'MG',
 'MG',
 'MG',
 'MG',
 'CL',
 'CL',
 'MG',
 'NA',
 'CL',
 'CL',
 'CL',
 'CL',
 'BA',
 'CL',
 'CL',
 'CL',
 'CL',
 'CL',
 'CL',
 'CL',
 'CL',
 'MG',
 'MG',
 'MG',
 'CL',
 'CL',
 'NA',
 'NA',
 'CL',
 'CL',
 'NI',
 'NI',
 'NI',
 'NI',
 'NI',
 'NI',
 'NI',
 'NI',
 'NI',
 'CA',
 'CL',
 'CL',
 'MG',
 'CL',
 'CL',
 'CL',
 'CL',
 'MG',
 'CL',

In [9]:
set(residues)

{'AG',
 'AL',
 'AM',
 'AR',
 'AU',
 'BA',
 'BR',
 'C08',
 'C10',
 'C11',
 'C14',
 'C15',
 'C19',
 'C1B',
 'C20',
 'C24',
 'C2G',
 'C35',
 'C44',
 'CA',
 'CA3',
 'CAK',
 'CD',
 'CE',
 'CL',
 'CO',
 'CR',
 'CS',
 'CU',
 'DY',
 'EU',
 'F',
 'F9F',
 'FE',
 'GA',
 'GD',
 'HDM',
 'HG',
 'HO',
 'I',
 'IN',
 'IR',
 'K',
 'KR',
 'LA',
 'LI',
 'LU',
 'MG',
 'MN',
 'MO',
 'MO7',
 'N25',
 'NA',
 'NI',
 'O',
 'O28',
 'OAN',
 'OS',
 'P',
 'PB',
 'PD',
 'PR',
 'PT',
 'RB',
 'RE',
 'RU',
 'SB',
 'SE',
 'SM',
 'SR',
 'TB',
 'TE',
 'TH',
 'TL',
 'UNL',
 'V',
 'W',
 'XE',
 'YB',
 'ZN'}

In [10]:
residues_metals = list(set(residues))

In [11]:
residues_metals.remove('AR')

In [12]:
residues_metals.remove('BR')

In [13]:
residues_metals.remove('XE')

In [14]:
metals_list = ['TH',
 'BA',
 'YB',
 'EU',
 'FE',
 'DY',
 'V',
 'GA',
 'HG',
 'PR',
 'NI',
 'PT',
 'LA',
 'NA',
 'LI',
 'PB',
 'RE',
 'TL',
 'LU',
 'RU',
 'RB',
 'TE',
 'TB',
 'K',
 'ZN',
 'CO',
 'PD',
 'AG',
 'CA',
 'IR',
 'AM',
 'AL',
 'CE',
 'CD',
 'GD',
 'AU',
 'W',
 'IN',
 'CS',
 'CR',
 'CU',
 'MG',
 'SR',
 'MO',
 'MN',
 'SM',
 'SB',
 'OS',
 'HO']

In [15]:
residues_metals = metals_list

In [16]:
counter = {}

for i in residues_metals:
    counter[i] = 0

In [17]:
for i in residues:
    for j in counter:
        if i == j:
            counter[j] += 1

In [18]:
counter

{'AG': 84,
 'AL': 10,
 'AM': 6,
 'AU': 241,
 'BA': 626,
 'CA': 26816,
 'CD': 4525,
 'CE': 7,
 'CO': 1767,
 'CR': 15,
 'CS': 544,
 'CU': 3544,
 'DY': 26,
 'EU': 33,
 'FE': 4028,
 'GA': 7,
 'GD': 181,
 'HG': 1613,
 'HO': 34,
 'IN': 1,
 'IR': 7,
 'K': 6681,
 'LA': 19,
 'LI': 87,
 'LU': 59,
 'MG': 56508,
 'MN': 9635,
 'MO': 115,
 'NA': 18313,
 'NI': 2298,
 'OS': 69,
 'PB': 189,
 'PD': 52,
 'PR': 59,
 'PT': 463,
 'RB': 197,
 'RE': 2,
 'RU': 59,
 'SB': 6,
 'SM': 122,
 'SR': 3869,
 'TB': 44,
 'TE': 7,
 'TH': 6,
 'TL': 400,
 'V': 1,
 'W': 172,
 'YB': 178,
 'ZN': 30094}

In [19]:
from __future__ import division

In [20]:
counter_perc = {}

total = 0

for i in counter:
    total += counter[i]
    
for i in counter:
    counter_perc[i] = counter[i] / total * 100
    
counter_perc    

{'AG': 0.04832613235607155,
 'AL': 0.005753110994770422,
 'AM': 0.0034518665968622532,
 'AU': 0.13864997497396717,
 'BA': 0.36014474827262843,
 'CA': 15.427542443576364,
 'CD': 2.6032827251336164,
 'CE': 0.004027177696339296,
 'CO': 1.0165747127759335,
 'CR': 0.008629666492155633,
 'CS': 0.312969238115511,
 'CU': 2.0389025365466376,
 'DY': 0.014958088586403096,
 'EU': 0.018985266282742393,
 'FE': 2.317353108693526,
 'GA': 0.004027177696339296,
 'GD': 0.10413130900534463,
 'HG': 0.927976803456469,
 'HO': 0.019560577382219436,
 'IN': 0.0005753110994770422,
 'IR': 0.004027177696339296,
 'K': 3.8436534556061193,
 'LA': 0.010930910890063802,
 'LI': 0.050052065654502674,
 'LU': 0.03394335486914549,
 'MG': 32.50967960924871,
 'MN': 5.543122443461302,
 'MO': 0.06616077643985985,
 'NA': 10.535672164723074,
 'NI': 1.322064906598243,
 'OS': 0.03969646586391591,
 'PB': 0.10873379780116096,
 'PD': 0.02991617717280619,
 'PR': 0.03394335486914549,
 'PT': 0.26636903905787057,
 'RB': 0.1133362865969773

In [21]:
sorted_perc = sorted(counter_perc.items(), key = lambda x: x[1], reverse = True)

In [22]:
sorted_perc

[('MG', 32.50967960924871),
 ('ZN', 17.31341222766211),
 ('CA', 15.427542443576364),
 ('NA', 10.535672164723074),
 ('MN', 5.543122443461302),
 ('K', 3.8436534556061193),
 ('CD', 2.6032827251336164),
 ('FE', 2.317353108693526),
 ('SR', 2.225878643876676),
 ('CU', 2.0389025365466376),
 ('NI', 1.322064906598243),
 ('CO', 1.0165747127759335),
 ('HG', 0.927976803456469),
 ('BA', 0.36014474827262843),
 ('CS', 0.312969238115511),
 ('PT', 0.26636903905787057),
 ('TL', 0.23012443979081687),
 ('AU', 0.13864997497396717),
 ('RB', 0.11333628659697731),
 ('PB', 0.10873379780116096),
 ('GD', 0.10413130900534463),
 ('YB', 0.1024053757069135),
 ('W', 0.09895350911005127),
 ('SM', 0.07018795413619915),
 ('MO', 0.06616077643985985),
 ('LI', 0.050052065654502674),
 ('AG', 0.04832613235607155),
 ('OS', 0.03969646586391591),
 ('PR', 0.03394335486914549),
 ('LU', 0.03394335486914549),
 ('RU', 0.03394335486914549),
 ('PD', 0.02991617717280619),
 ('TB', 0.025313688376989857),
 ('HO', 0.019560577382219436),
 (

In [26]:
sorted_perc_big = []

for i in sorted_perc:
    if i[1] > 1:
        sorted_perc_big.append(i)

In [27]:
sorted_perc_big

[('MG', 32.50967960924871),
 ('ZN', 17.31341222766211),
 ('CA', 15.427542443576364),
 ('NA', 10.535672164723074),
 ('MN', 5.543122443461302),
 ('K', 3.8436534556061193),
 ('CD', 2.6032827251336164),
 ('FE', 2.317353108693526),
 ('SR', 2.225878643876676),
 ('CU', 2.0389025365466376),
 ('NI', 1.322064906598243),
 ('CO', 1.0165747127759335)]

In [28]:
total_big = 0

for i in sorted_perc_big:
    total_big += i[1]
    
total_big    

96.69713897790231

In [30]:
sorted_perc_big.append(('Others', 100 - total_big))

In [31]:
sorted_perc_big

[('MG', 32.50967960924871),
 ('ZN', 17.31341222766211),
 ('CA', 15.427542443576364),
 ('NA', 10.535672164723074),
 ('MN', 5.543122443461302),
 ('K', 3.8436534556061193),
 ('CD', 2.6032827251336164),
 ('FE', 2.317353108693526),
 ('SR', 2.225878643876676),
 ('CU', 2.0389025365466376),
 ('NI', 1.322064906598243),
 ('CO', 1.0165747127759335),
 ('Others', 3.3028610220976873)]

In [86]:
sorted_perc_big_corr = [(), (), (), (), (), (), (), (), (), (), (), (), ()]
sorted_perc_big_corr[0] = ('Mg', sorted_perc_big[0][1])
sorted_perc_big_corr[1] = ('Zn', sorted_perc_big[1][1])
sorted_perc_big_corr[2] = ('Ca', sorted_perc_big[2][1])
sorted_perc_big_corr[3] = ('Na', sorted_perc_big[3][1])
sorted_perc_big_corr[4] = ('Mn', sorted_perc_big[4][1])
sorted_perc_big_corr[5] = ('K', sorted_perc_big[5][1])
sorted_perc_big_corr[6] = ('Cd', sorted_perc_big[6][1])
sorted_perc_big_corr[7] = ('Fe', sorted_perc_big[7][1])
sorted_perc_big_corr[8] = ('Sr', sorted_perc_big[8][1])
sorted_perc_big_corr[9] = ('Cu', sorted_perc_big[9][1])
sorted_perc_big_corr[10] = ('Ni', sorted_perc_big[10][1])
sorted_perc_big_corr[11] = ('Co', sorted_perc_big[11][1])
sorted_perc_big_corr[12] = ('Others', sorted_perc_big[12][1])

In [88]:
sorted_perc_big_corr

[('Mg', 32.50967960924871),
 ('Zn', 17.31341222766211),
 ('Ca', 15.427542443576364),
 ('Na', 10.535672164723074),
 ('Mn', 5.543122443461302),
 ('K', 3.8436534556061193),
 ('Cd', 2.6032827251336164),
 ('Fe', 2.317353108693526),
 ('Sr', 2.225878643876676),
 ('Cu', 2.0389025365466376),
 ('Ni', 1.322064906598243),
 ('Co', 1.0165747127759335),
 ('Others', 3.3028610220976873)]

[('Mg', 32.50967960924871),
 ('Zn', 17.31341222766211),
 ('Ca', 15.427542443576364),
 ('Na', 10.535672164723074),
 ('Mn', 5.543122443461302),
 ('K', 3.8436534556061193),
 ('Cd', 2.6032827251336164),
 ('Fe', 2.317353108693526),
 ('Sr', 2.225878643876676),
 ('Cu', 2.0389025365466376),
 ('Ni', 1.322064906598243),
 ('Co', 1.0165747127759335),
 ('Others', 3.3028610220976873)]

In [23]:
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

In [80]:
matplotlib.rcParams['font.size'] = 16.0

In [108]:
explode = (0.2, 0.2, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15)
#colors = ['#1c7ce3', '#9524db', '#37c898', '#ffaea8', '#f5e50a', '#e8179f', '#36c991', '#1100ff', '#6e8d91', '#abed12', '#db1d1a', '#dd4e22', '#01e9fe']
#colors = ['#114B5F', '#FF4E00', '#028090', '#E4FDE1', '#8EA604', '#456990', '#F5BB00', '#F45B69', '#EC9F05', '#BF3100']
colors = ['#FF9F1C', '#028090', '#37c898', '#ffaea8', '#f5e50a', '#2EC4B6', '#8EA604', '#7FB800', '#00A6ED', '#BF3100', '#F6511D']
plt.figure(figsize = (11,11))
plt.pie([i[1] for i in sorted_perc_big_corr], labels = [i[0] for i in sorted_perc_big_corr], autopct='%1.1f%%', pctdistance = 0.8, labeldistance = 1.1, colors = colors, explode = explode, shadow = True)
plt.axis('equal')
plt.title('Metals in the PDB - % of all metal atoms (in 33% of structures)')
plt.savefig('major_metals.png')

In [109]:
perc_others = []

for i in sorted_perc:
    if i[1] <= 1:
        perc_others.append(i)

In [110]:
perc_others

[('HG', 0.927976803456469),
 ('BA', 0.36014474827262843),
 ('CS', 0.312969238115511),
 ('PT', 0.26636903905787057),
 ('TL', 0.23012443979081687),
 ('AU', 0.13864997497396717),
 ('RB', 0.11333628659697731),
 ('PB', 0.10873379780116096),
 ('GD', 0.10413130900534463),
 ('YB', 0.1024053757069135),
 ('W', 0.09895350911005127),
 ('SM', 0.07018795413619915),
 ('MO', 0.06616077643985985),
 ('LI', 0.050052065654502674),
 ('AG', 0.04832613235607155),
 ('OS', 0.03969646586391591),
 ('PR', 0.03394335486914549),
 ('LU', 0.03394335486914549),
 ('RU', 0.03394335486914549),
 ('PD', 0.02991617717280619),
 ('TB', 0.025313688376989857),
 ('HO', 0.019560577382219436),
 ('EU', 0.018985266282742393),
 ('DY', 0.014958088586403096),
 ('LA', 0.010930910890063802),
 ('CR', 0.008629666492155633),
 ('AL', 0.005753110994770422),
 ('GA', 0.004027177696339296),
 ('TE', 0.004027177696339296),
 ('IR', 0.004027177696339296),
 ('CE', 0.004027177696339296),
 ('TH', 0.0034518665968622532),
 ('AM', 0.0034518665968622532),


In [112]:
others_total = 0

for i in perc_others:
    others_total += i[1]
    
others_total, 100 - others_total

(3.3028610220976997, 96.6971389779023)

In [113]:
100 / others_total

30.2767810485978

In [114]:
perc_others_mult = []

for i in perc_others:
    perc_others_mult.append((i[0], i[1] * (100/others_total)))
    
perc_others_mult    

[('HG', 28.09615049642919),
 ('BA', 10.904023689252742),
 ('CS', 9.475701097369795),
 ('PT', 8.064797073680543),
 ('TL', 6.96742727747779),
 ('AU', 4.197874934680368),
 ('RB', 3.4314579341578115),
 ('PB', 3.2921093886082557),
 ('GD', 3.1527608430587),
 ('YB', 3.1005051384776166),
 ('W', 2.99599372931545),
 ('SM', 2.125065319630726),
 ('MO', 2.0031353422748643),
 ('LI', 1.5154154328514196),
 ('AG', 1.4631597282703361),
 ('OS', 1.2018812053649188),
 ('PR', 1.0276955234279739),
 ('LU', 1.0276955234279739),
 ('RU', 1.0276955234279739),
 ('PD', 0.9057655460721127),
 ('TB', 0.7664170005225569),
 ('HO', 0.5922313185856122),
 ('EU', 0.5748127503919177),
 ('DY', 0.45288277303605634),
 ('LA', 0.33095279568019503),
 ('CR', 0.26127852290541714),
 ('AL', 0.17418568193694475),
 ('GA', 0.12192997735586135),
 ('TE', 0.12192997735586135),
 ('IR', 0.12192997735586135),
 ('CE', 0.12192997735586135),
 ('TH', 0.10451140916216685),
 ('AM', 0.10451140916216685),
 ('SB', 0.10451140916216685),
 ('RE', 0.034837

In [116]:
new_total = 0

for i in perc_others_mult:
    new_total += i[1]

In [117]:
new_total

99.99999999999999

## FILES analysis

In [53]:
ls

17Nov_FILES_metal_searcher_results.txt  17Nov_metal_searcher_analysis.ipynb     metal_searcher_results.txt              metals_list.txt


In [54]:
results_files = open('17Nov_FILES_metal_searcher_results.txt')

In [55]:
files_string = results_files.read()

In [56]:
files_string

"{'metals_in_file': {'BA': 128, 'W': 14, 'YB': 73, 'CS': 100, 'FE': 1250, 'DY': 1, 'HG': 426, 'PR': 25, 'NI': 946, 'PT': 91, 'LA': 8, 'NA': 5462, 'LI': 55, 'PB': 45, 'RE': 2, 'TL': 54, 'LU': 14, 'RB': 44, 'IN': 1, 'TE': 3, 'TB': 13, 'CA': 8339, 'PD': 12, 'ZN': 10407, 'CO': 603, 'TH': 2, 'AG': 19, 'RU': 20, 'K': 1891, 'IR': 3, 'AM': 1, 'AL': 4, 'CE': 4, 'CD': 803, 'GD': 38, 'AU': 57, 'GA': 6, 'V': 1, 'EU': 11, 'CR': 8, 'CU': 1020, 'MG': 10021, 'SR': 126, 'MO': 31, 'MN': 2418, 'SM': 36, 'SB': 3, 'OS': 11, 'HO': 11}, 'any_metal_in_file': 37292, 'ok_file_count': 113129, 'error_file_count': 195}"

In [96]:
37292 / 113129

0.3296413828461314